### PR TITLE
Flux monitor multinode partial integration test

### PR DESCRIPTION
**Pull request by coventry**

### CODE REVIEW

File: core/internal/flux_monitor_simulated_blockchain_test.go [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+	fa                          *fluxAggregator
+	roundId, answer             int64
+	from                        *bind.TransactOpts
+	isNewRound, completesAnswer bool
+}
+
+// checkUpdateAnswer verifies all the logs emitted by fa's FluxAggregator
+// contract after an updateAnswer with the given values.
+func checkUpdateAnswer(t *testing.T, p answerParams,
+	currentBalance int64, receiptBlock uint64) {
```
Comment (spooktheducks) [inline core/internal/flux_monitor_simulated_blockchain_test.go:150]: Might want to add a call to `t.Helper()` at the top of this method.

File: core/internal/flux_monitor_simulated_blockchain_test.go [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+		t.Fatalf("chainlink failed to submit answer to FluxAggregator contract")
+	}
+	checkUpdateAnswer(t, answerParams{fa: &fa, roundId: roundId,
+		answer: processedAnswer, from: fa.nallory, isNewRound: false,
+		completesAnswer: true}, initialBalance, receiptBlock)
+
+	//- have the malicious node start the next round.
+	nextRoundBalance := initialBalance - fee
+	// Triggers a new round, since price deviation exceeds threshold
+	reportPrice = answer + 1
```
Comment (spooktheducks) [inline core/internal/flux_monitor_simulated_blockchain_test.go:310]: Is there any chance we could get a data race leading to intermittent test failures here?

File: core/internal/flux_monitor_simulated_blockchain_test.go [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+	//- successfully close the round through the submissions of the other nodes
+	updateAnswer(t, answerParams{fa: &fa, roundId: newRound,
+		answer: processedAnswer, from: fa.neil, isNewRound: false,
+		completesAnswer: true})
+
+	//- have the malicious node try to start another round repeatedly until the
+	//roundDelay is reached, making sure that it isn't successful
+	// Triggers a new round, since price deviation exceeds threshold
+	reportPrice = answer + 1
+	select {
```
Comment (spooktheducks) [inline core/internal/flux_monitor_simulated_blockchain_test.go:331]: I think you might be able to do something like `require.Consistently(t, func() bool { return len(submissionReceived) == 0 })`.  You would need to make it a buffered channel of size 1, though.

File: core/internal/flux_monitor_simulated_blockchain_test.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+	case <-submissionReceived:
+		t.Fatalf("chainlink node updated FA, even though it's not allowed to")
+	case <-time.After(5 * timeout):
+	}
+	// Could add a check for "not eligible to submit here", using the memory log
```
Comment (spooktheducks) [inline core/internal/flux_monitor_simulated_blockchain_test.go:336]: This sounds like a good idea -- want to make sure it's failing to submit for the right reason.

File: core/internal/gethwrappers/generation/generate.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+///bin/true; exec /usr/bin/env go run "$0" "$@"
```
Comment (spooktheducks) [inline core/internal/gethwrappers/generation/generate.go:1]: This is awesome.

File: core/services/fluxmonitor/flux_monitor.go [hunk 6]
```diff
@@ -198,14 +200,27 @@ func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
...
 	if id == nil {
 		logger.Warn("nil job ID passed to FluxMonitor#RemoveJob")
 		return
 	}
 	fm.chRemove <- *id
 }
 
+// XXXTestingOnlyCreateJob creates a job with a specific answer and round, for
+// testing nodes with malicious behavior
+func (fm *ConcreteFluxMonitor) XXXTestingOnlyCreateJob(t *testing.T,
```
Comment (spooktheducks) [inline core/services/fluxmonitor/flux_monitor.go:213]: There's actually a better way to do this.  If you create a file called, say, `helpers_test.go`, and then at the top of the file, declare it to be a member of `package fluxmonitor`, you can put this declaration in there, and it will be available to your tests.  When the application is built normally, that file will not be included, and so the method won't be available.  We should already have a `helpers_test.go` in the FM package, come to think of it.  Check that out for some examples.

File: core/services/fluxmonitor/flux_monitor.go [hunk 7]
```diff
@@ -639,9 +656,13 @@ var (
 func (p *PollingDeviationChecker) checkEligibilityAndAggregatorFunding(roundState contracts.FluxAggregatorRoundState) error {
 	if !roundState.EligibleToSubmit {
 		return ErrNotEligible
-	} else if roundState.AvailableFunds.Cmp(roundState.PaymentAmount) < 0 {
+	}
+	if roundState.AvailableFunds.Cmp(roundState.PaymentAmount) < 0 {
```
Comment (spooktheducks) [inline core/services/fluxmonitor/flux_monitor.go:660]: Why the change?

File: core/internal/cltest/client.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+	number interface{}) (blockNumber *big.Int, err error) {
+	switch n := number.(type) {
+	case string:
+		switch n {
+		case "latest":
+			return c.currentBlockNumber(), nil
+		case "earliest":
+			return big.NewInt(0), nil
+		case "pending":
+			panic("not implemented") // I don't understand the semantics of this.
```
[the remark this answers is not in the document]
Comment (coventry) [reply core/internal/cltest/client.go:256]: Yeah, this is confusing to me. `c.currentBlockNumber()`, which is used for `latest`, seems to be the block that is now being generated. That's why `GetBlockHeight` subtracts 1 from it, returning the last *committed* block. So what this is calling `latest`, I would have called "pending". It seems to me that there is no legitimate use for the `pending` option.

File: core/internal/flux_monitor_simulated_blockchain_test.go (hunk 3 above)
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go:331]: Do you mean something like 

```golang
	g := gomega.NewGomegaWithT(t)
	g.Eventually(submissionReceived).Should(gomega.Receive(),
		"chainlink failed to submit answer to FluxAggregator contract")
```
?

This is failing with assertion failures in `checkUpdateAnswer`:

```
    flux_monitor_simulated_blockchain_test.go:166: 
        	Error Trace:	flux_monitor_simulated_blockchain_test.go:166
        	            				flux_monitor_simulated_blockchain_test.go:317
        	Error:      	Should be true
        	Test:       	TestFluxMonitorAntiSpamLogic
        	Messages:   	SubmissionReceived log has wrong answer
    flux_monitor_simulated_blockchain_test.go:168: 
        	Error Trace:	flux_monitor_simulated_blockchain_test.go:168
        	            				flux_monitor_simulated_blockchain_test.go:317
        	Error:      	Not equal: 
        	            	expected: 0x2
        	            	actual  : 0x1
        	Test:       	TestFluxMonitorAntiSpamLogic
        	Messages:   	SubmissionReceived log has wrong round
    flux_monitor_simulated_blockchain_test.go:178: 
        	Error Trace:	flux_monitor_simulated_blockchain_test.go:178
        	            				flux_monitor_simulated_blockchain_test.go:317
        	Error:      	"[]" should have 1 item(s), but has 0
        	Test:       	TestFluxMonitorAntiSpamLogic
        	Messages:   	FluxAggregator did not emit correct NewRound log
```

I think the issue may be that [`Receive`](http://onsi.github.io/gomega/#receive) doesn't block.

File: core/internal/gethwrappers/generation/generate.go (hunk 5 above)
Comment (coventry) [reply core/internal/gethwrappers/generation/generate.go:1]: Oh, damn, I didn't mean to check that in. Still, it's cool, and means we can run go with `go generate` commands.

File: core/services/fluxmonitor/flux_monitor.go (hunk 6 above)
Comment (coventry) [reply core/services/fluxmonitor/flux_monitor.go:213]: I believe it will  only compile `*_test.go` files in the package under test. I tried it here, and got a compiler error. `helpers_test.go` symbols are only referenced in the `fluxmonitor` package.

File: core/services/fluxmonitor/flux_monitor.go (hunk 7 above)
Comment (coventry) [reply core/services/fluxmonitor/flux_monitor.go:660]: I needed it for debugging. Gratuitous `else if` chains should be cleansed with fire, because they make the code inflexible. I'll remove this, though.

File: core/internal/flux_monitor_simulated_blockchain_test.go (hunk 4 above)
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go:336]: Added. (I was able to do this with `gomega.Eventually`.)

File: core/internal/cltest/mocks.go [hunk 9]
```diff
@@ -595,6 +595,16 @@ func NewHTTPMockServer(
 	}
 }
 
+func NewHTTPMockServerWithAlterableResponse(
```
Comment (se3000) [inline core/internal/cltest/mocks.go:598]: 👍

File: core/internal/gethwrappers/generation/compile_command.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+yarn && myarn workspace @chainlink/contracts belt compile solc
```
Comment (se3000) [inline core/internal/gethwrappers/generation/compile_command.txt:1]: What is `myarn`? Typo? If so, is this file being used?

File: core/services/fluxmonitor/flux_monitor.go [hunk 11]
```diff
@@ -45,7 +47,7 @@ type Service interface {
 	Stop()
 }
 
-type concreteFluxMonitor struct {
+type ConcreteFluxMonitor struct {
```
Comment (se3000) [inline core/services/fluxmonitor/flux_monitor.go:50]: Why is this exported now?

File: core/internal/gethwrappers/generation/compile_command.txt (hunk 10 above)
Comment (coventry) [reply core/internal/gethwrappers/generation/compile_command.txt:1]: Yes, it was a typo, and yes, this file is used. Thanks for catching it.

File: core/internal/cltest/client.go [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+// Subscribe is a dummy method present only to satisfy the eth.Client interface.
+// The original method is for subscribing to events observed by the RPC client,
+// but for a simulated backend that makes no sense.
+func (c *SimulatedBackendClient) Subscribe(ctx context.Context, namespace interface{},
+	channelAndArgs ...interface{}) (eth.Subscription, error) {
+	// if these are needed, there are subscribe methods on SimulatedBackend
+	panic("unimplemented")
+}
+
+// XXX: Move these to utils.
```
[the remark this answers is not in the document]
Comment (coventry) [reply core/internal/cltest/client.go:106]: I moved it to `services/eth`. `utils` created an import cycle.

File: core/internal/flux_monitor_simulated_blockchain_test.go [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+	initr.InitiatorParams.PollingInterval =
+		models.Duration(timeout)
+	initr.InitiatorParams.Address = fa.aggregatorContractAddress
+	j := cltest.CreateJobSpecViaWeb(t, app, job)
+	jrs := cltest.WaitForRuns(t, j, app.Store, 1) // Submit answer from
+	reportedPrice := jrs[0].RunRequest.RequestParams.Get("result").String()
+	assert.Equal(t, reportedPrice, fmt.Sprintf("%d", reportPrice),
+		"failed to report correct price to contract")
+	var receiptBlock uint64
+	select { // block until FluxAggregator contract acknowledges chainlink message
```
[the remark this answers is not in the document]
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go]: I tried this, and because I would still need a `switch` statement to get a non-blocking read in the argument to `gomega.Eventually`, I decided it's clearer the way it is. Please LMK if I've missed something.

[the remark this answers is not in the document]
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go]: I don't really understand `t.Helper`'s purpose. Why would I want to elide information from test failure output?

File: core/internal/flux_monitor_simulated_blockchain_test.go (hunk 1 above)
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go]: I don't really understand `t.Helper`'s purpose. Why would I want to elide information from test failure output?

File: core/internal/flux_monitor_simulated_blockchain_test.go (hunk 2 above)
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go]: No, that's the role of the `select`s on either side of this instruction. They will wait for the node to respond to the reported change before allowing the test to continue (or time out and throw an error.)

File: core/internal/cltest/client.go [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+	}
+	callArgs, ok := args[0].(eth.CallArgs)
+	if !ok {
+		return nil, nil, fmt.Errorf("third arg to SimulatedBackendClient.Call "+
+			"must be an eth.CallArgs, got %+#v", args[0])
+	}
+	blockNumber, err := c.blockNumber(args[1])
+	if err != nil || blockNumber.Cmp(c.currentBlockNumber()) != 0 {
+		return nil, nil, fmt.Errorf("fourth arg to SimulatedBackendClient.Call "+
+			"must the string \"latest\", or a *big.Int equal to current "+
```
Comment (RyanRHall) [inline core/internal/cltest/client.go:56]: "must be the string"

File: core/internal/gethwrappers/go_generate_test.go [hunk 15]
```diff
@@ -206,10 +206,34 @@ func init() {
...
+		if !whitespace([]byte{c}) || inQuotes {
+			rv = append(rv, c)
+		}
+	}
+	return string(rv)
+}
+
+func ExamplestripWhitespaceOutsideQuotes() {
+	fmt.Print(stripWhitespaceOutsideQuotes(
+		`The rain in spain stays mainly on "the plain"`))
```
Comment (RyanRHall) [inline core/internal/gethwrappers/go_generate_test.go:233]: :+1:

File: core/internal/cltest/client.go (hunk 14 above)
Comment (coventry) [reply core/internal/cltest/client.go:56]: Oops, missed this before merging. Thanks.

**Comment by coventry** [comment]
Thanks, @RyanRHall !

File: core/internal/flux_monitor_simulated_blockchain_test.go (hunk 1 above)
Comment (coventry) [reply core/internal/flux_monitor_simulated_blockchain_test.go]: I added this, although I still don't really understand its purpose.

File: core/eth/types.go [hunk 16]
```diff
@@ -290,3 +290,13 @@ func (ary UntrustedBytes) SafeByteSlice(start int, end int) ([]byte, error) {
 	}
 	return ary[start:end], nil
 }
+
+// ChainlinkEthLogFromGethLog returns a copy of l as an eth.Log. (They have
+// identical fields, but the field tags differ, and the types differ slightly.)
+//
+// A better place for this might be utils, but that creates an import cycle
+func ChainlinkEthLogFromGethLog(l types.Log) Log {
```
Comment (samsondav) [inline core/eth/types.go:298]: Pretty sure I already wrote something identical to this already... try `toCLEthLog`

File: core/internal/cltest/client.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/internal/cltest/client.go:250]: This comment is resolved but I dont understand what the resolution was.

File: core/internal/cltest/cltest.go [hunk 17]
```diff
@@ -99,26 +103,32 @@ func NewConfig(t testing.TB) (*TestConfig, func()) {
 	return NewConfigWithWSServer(t, wsserver), cleanup
 }
 
+var testConfigConstants map[string]interface{} = map[string]interface{}{
```
Comment (samsondav) [inline core/internal/cltest/cltest.go:106]: Why has this moved? This is going to give me mad cornflakes :(

File: core/internal/cltest/cltest.go [hunk 18]
```diff
@@ -246,6 +257,25 @@ func NewApplicationWithConfig(t testing.TB, tc *TestConfig, flags ...string) (*T
...
+	app, appCleanup := NewApplicationWithConfigAndKey(t, tc, flags...)
+	var client SimulatedBackendClient
+	if txm, ok := app.Store.TxManager.(*store.EthTxManager); ok {
+		client = SimulatedBackendClient{b: backend, t: t, chainId: chainId}
+		txm.Client = &client
+	} else {
+		log.Panic("SimulatedBackend only works on EthTxManager")
+	}
+	// Clean out the mock registrations, since we don't need those...
+	app.EthMock.Responses = app.EthMock.Responses[:0]
```
Comment (samsondav) [inline core/internal/cltest/cltest.go:274]: urgh... I wonder if we could avoid mocking the eth client entirely if this is the case?

File: core/internal/cltest/cltest.go [hunk 19]
```diff
@@ -1206,3 +1236,32 @@ func MakeRoundStateReturnData(
...
+// EthereumLogIterator is the interface provided by gethwrapper representations of EVM
+// logs.
+type EthereumLogIterator interface{ Next() bool }
+
+// GetLogs drains logs of EVM log representations. Since those log
+// representations don't fit into a type hierarchy, this API is a bit awkward.
+// It returns the logs as a slice of blank interface{}s, and if rv is non-nil,
+// it must be a pointer to a slice for elements of the same type as the logs,
+// in which case GetLogs will append the logs to it.
+func GetLogs(t *testing.T, rv interface{}, logs EthereumLogIterator) []interface{} {
```
Comment (samsondav) [inline core/internal/cltest/cltest.go:1249]: Will the new generics fix this?

File: core/internal/gethwrappers/go_generate_test.go [hunk 20]
```diff
@@ -1,6 +1,6 @@
 // package vrf_test verifies correct and up-to-date generation of golang wrappers
 // for solidity contracts. See go_generate.go for the actual generation.
-package vrf_test
+package gethwrappers
```
Comment (samsondav) [inline core/internal/gethwrappers/go_generate_test.go:3]: why not `gethwrappers_test`?

File: core/internal/gethwrappers/go_generate_test.go [hunk 21]
```diff
@@ -206,10 +206,34 @@ func init() {
...
 var (
-	stripWhitespace    = regexp.MustCompile(`\s+`).ReplaceAllString
 	stripTrailingColon = regexp.MustCompile(":$").ReplaceAllString
+	whitespace         = regexp.MustCompile(`\s`).Match
+	quote              = []byte(`"`)[0]
 )
 
+// stripWhitespaceOutsideQuotes removes all whitespace outside " " quotes. Used
+// to normalize JSON for comparison in a hash. Fortunately, no quoted quotes
+// occur in the input
```
Comment (samsondav) [inline core/internal/gethwrappers/go_generate_test.go:216]: This feels dangerously hacky, can we not use a JSON library to do this for us?

Comment (samsondav) [reply core/internal/gethwrappers/go_generate_test.go:216]: Or `assert.JSONEq`?

File: core/services/fluxmonitor/flux_monitor.go [hunk 22]
```diff
@@ -989,7 +989,6 @@ func OutsideDeviation(curAnswer, nextAnswer decimal.Decimal, threshold float64)
 // If system time is not accurate (compared to the cluster) then you should
 // expect poor behaviour here.
 func MakeIdleTimer(log contracts.LogNewRound, idleThreshold models.Duration, clock utils.AfterNower) <-chan time.Time {
-	fmt.Println("log.StartedAt", log.StartedAt)
 	timeNow := clock.Now()
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go:992]: Thank you 🤠

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 23]
```diff
@@ -130,7 +130,7 @@ contract FluxAggregator is AggregatorInterface, Owned {
    * @notice Deploy with the address of the LINK token and initial payment amount
    * @dev Sets the LinkToken address and amount of LINK paid
    * @param _link The address of the LINK token
-   * @param _paymentAmount The amount paid of LINK paid to each oracle per submission
+   * @param _paymentAmount The amount paid of LINK paid to each oracle per submission, in wei (units of 10⁻¹⁸ LINK)
```
Comment (samsondav) [inline evm-contracts/src/v0.6/dev/FluxAggregator.sol:133]: Is that unicode? 🧐

File: core/internal/cltest/cltest.go (hunk 17 above)
Comment (coventry) [reply core/internal/cltest/cltest.go:106]: What are cornflakes?

File: core/internal/gethwrappers/go_generate_test.go (hunk 21 above)
Comment (coventry) [reply core/internal/gethwrappers/go_generate_test.go:216]: Currently, the DB tracks hashes of the JSON. Storing the full JSON would be necessary to use `JSONEq`. Essentially, that would be duplicating all the ABI JSON used in `gethwrappers`.

It's hacky, but it's hacky in a way which complains loudly when it breaks: The hashes won't match.

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol (hunk 23 above)
Comment (coventry) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol:133]: Yep. Unicode is valid in golang. I use it a fair bit in the VRF code, too.

File: core/internal/cltest/cltest.go (hunk 17 above)
Comment (coventry) [reply core/internal/cltest/cltest.go:106]: This was broken out due to an earlier strategy I had for managing the chain ID. I can revert this change.

File: core/internal/gethwrappers/go_generate_test.go (hunk 21 above)
Comment (j16r) [reply core/internal/gethwrappers/go_generate_test.go:216]: You should be using `NormalizedJSON` surely.

Comment (coventry) [reply core/internal/gethwrappers/go_generate_test.go:216]: You're right, I should. Thanks.

File: core/eth/types.go (hunk 16 above)
Comment (coventry) [reply core/eth/types.go:298]: I moved this into `cltest.go`. That triggered an import cycle in `vrf`, so I moved all the vrf test code to `vrf_test`, and there are a few dozen minor  changes resulting from that.

**Comment by coventry** [comment]
> This comment is resolved but I dont understand what the resolution was.

@samsondav , I don't think we're going to come to a full resolution on this. `latest` is definitely what we need, and `pending` semantics in a simulated blockchain is unclear to me at this stage. Since we have no need for  `pending`, at this point, I'm fine leaving it unimplemented, at least for now.

File: core/internal/gethwrappers/go_generate_test.go (hunk 21 above)
Comment (coventry) [reply core/internal/gethwrappers/go_generate_test.go:216]: I can't believe that just worked, even though I'm comparing to `jq` output....

File: core/internal/cltest/cltest.go (hunk 18 above)
Comment (coventry) [reply core/internal/cltest/cltest.go:269]: I guess we could, but the point of a mock is it's cheap, and I don't see much potential for confusion by keeping this around. I think at least this should go in a future story.

**Comment by samsondav** [comment]
@coventry golang/static Shadow failing

**Comment by coventry** [comment]
@samsondav , yeah, and thanks for taking a look. 

The output of that test is completely uninformative. I asked about this in #engineering, and received no response, but while I was responding to you, I figured out how to reproduce the test locally:

```bash
(cd /tmp; go install golang.org/x/tools/go/analysis/passes/shadow/cmd/shadow)
go vet -vettool=$(which shadow) ./...
```

I've pushed a fix.

File: core/services/fluxmonitor/helpers_test.go [hunk 24]
```diff
@@ -85,3 +88,17 @@ func dataWithResult(t *testing.T, result decimal.Decimal) adapterResponseData {
 	require.NoError(t, json.Unmarshal(body, &data))
 	return data
 }
+
+// XXXTestingOnlyCreateJob is used in TestFluxMonitorAntiSpamLogic to create a
+// job with a specific answer and round, for testing nodes with malicious
+// behavior
+func (fm *concreteFluxMonitor) XXXTestingOnlyCreateJob(t *testing.T,
```
Comment (se3000) [inline core/services/fluxmonitor/helpers_test.go:95]: This name is unnecessary now, given the method location.

File: core/internal/cltest/cltest.go (hunk 18 above)
Comment (se3000) [reply core/internal/cltest/cltest.go:269]: Agreed, would be nice to avoid the initial setup all together if possible. Just needless extra steps, could see it being very unintuitive when we change ethMock and the simulated backend tests start breaking.
